src/storage/src/wal.rs

```diff
@@ -0,0 +1,385 @@
+                current_bytes_written: 0,
+                total_bytes_written: 0,
+            };
+
+            Ok(ret)
+        } else {
+            bail!("couldn't find any in-progress log segments when trying to resume wal")
+        }
+    }
+}
+
+pub struct WriteAheadLogs {
+    path: PathBuf,
+    wals: HashMap<GlobalId, WriteAheadLog>,
+}
+
+impl WriteAheadLogs {
+    pub fn new(path: PathBuf) -> Result<Self, anyhow::Error> {
+        Ok(Self {
+            path,
+            wals: HashMap::new(),
+        })
+    }
+
+    // TODO make this path use cluster_id
+    fn get_path(&self, id: GlobalId) -> PathBuf {
+        self.path.join(id.to_string())
+    }
+
+    pub fn create(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if self.wals.contains_key(&id) {
+            bail!(
+                "asked to create wal for relation: {:?} that was already created",
+                id
+            )
+        }
+
+        let wal_base_path = self.get_path(id);
+        let wal = WriteAheadLog::create(id, wal_base_path)?;
+        self.wals.insert(id, wal);
+        Ok(())
+    }
+
+    pub fn destroy(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!(
+                "asked to delete wal for relation: {:?} that doesn't exist",
+                id
+            )
+        }
+
+        self.wals.remove(&id);
+        Ok(())
+    }
+
+    pub fn write(&mut self, id: GlobalId, updates: &[Update]) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!("asked to write to unknown relation: {}", id)
+        }
+
+        let wal = self.wals.get_mut(&id).expect("wal known to exist");
+        wal.write(updates)?;
+        Ok(())
+    }
+
+    pub fn resume(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if self.wals.contains_key(&id) {
+            bail!(
+                "asked to resume wal for relation: {:?} that already exists.",
+                id
+            )
+        }
+        let wal_path = self.get_path(id);
+        let wal = WriteAheadLog::resume(id, wal_path)?;
+        self.wals.insert(id, wal);
+
+        Ok(())
+    }
+
+    pub fn write_progress(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
+        // For all logs, indicate that <timestamp> is now the min timestamp going
+        // forward
+        for (_, wal) in self.wals.iter_mut() {
+            wal.write_progress(timestamp)?;
+        }
+
+        Ok(())
+    }
+}
+
+// Write a 20 byte header + length-prefixed Row to a buffer
+// TODO: change the on-disk representation to be more compact.
+pub fn encode_update(
+    row: &Row,
+    timestamp: Timestamp,
+    diff: isize,
+    buf: &mut Vec<u8>,
+) -> Result<(), anyhow::Error> {
+    assert!(diff != 0);
+    assert!(timestamp > 0);
+    let data = row.data();
+
+    if data.len() >= u32::MAX as usize {
+        bail!("failed to encode row: row too large");
```

Comment:
It seems to me that ```bail!``` causes the thread to exit. MZ remains up but all future updates are silently not persisted.

Reply:
this limit is fairly artificial and I can get rid of it.

Reply:
I think the limit is reasonable for the time being, there are various uses  of ```bail!``` and my comment was about all of them in general.

Reply:
One thing that I think would be really neat is to have persistence status visible as an introspection table - if we disable persisting in any way, we should make this visible to operators (outside of error displays), so they can alert, or better, so that we can get alerted.

Reply:
I think I've upgraded all erroring paths to print a log message saying something like "No further updates will be persisted" which is the most I think I can do in the context of this pr. I agree with philips suggestion that persistence failures should block future inserts instead of accepting them but that work will have to wait for the next pr I'm afraid.